### PR TITLE
kvc-simple-kmod implementation

### PR DESCRIPTION
**Pull request by zvonkok**
This demonstrated how to build a simple mod with KVC and the driver-container-base. 

```
# PULLPOLICY=Always make deploy 
# SPECIALRESOURCE=driver-container-base make specialresource
# SPECIALRESOURCE=simple-kmod make specialresource
```

### CODE REVIEW

File: deploy/crds/sro.openshift.io_specialresources_crd.yaml [hunk 1]
```diff
@@ -32,15 +32,19 @@ spec:
           description: SpecialResourceSpec defines the desired state of SpecialResource
           properties:
             dependsOn:
-              description: SpecialResourceDependsOn defines the desired state of SpecialResource
-              properties:
-                name:
-                  items:
+              items:
+                description: SpecialResourceDependency is a SpecialResource that needs
+                  to be Complete
```
Comment (kpouget) [inline deploy/crds/sro.openshift.io_specialresources_crd.yaml:37]: I'm not sure to understand this change, `SpecialResource|DependsOn` matches the name of the property, but `SpecialResource|Dependency` doesn't. Is the description wrong, or did you forget to change the name of the property?

Comment (kpouget) [reply deploy/crds/sro.openshift.io_specialresources_crd.yaml:37]: hum, reading the changes below, I see that `SpecialResourceDependency` is the name of the type of the objects inside `dependsOn` array, so maybe it's fine as it is now

File: recipes/simple-kmod/config/simple-kmod-cr.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+  dependsOn: 
+  - name: "driver-container-base"
+    namespace: "driver-container-base"
+
+    
+
+
+
+
```
Comment (kpouget) [inline recipes/simple-kmod/config/simple-kmod-cr.yaml]: empty lines can be removed

File: recipes/simple-kmod/config/Makefile [hunk 3]
```diff
@@ -0,0 +1,6 @@
+simple-kmod: 
+	-kubectl create ns $@
+	-kubectl delete configmap $@ -n $@  2>/dev/null | true
+	kubectl create configmap $@ -n $@ --from-file=recipes/$@/manifests
+	-oc policy add-role-to-user system:image-puller system:serviceaccount:$@:builder -n driver-container-base
+	@$(TEMPLATE_CMD) recipes/$@/config/$@-cr.yaml | kubectl apply -f -
```
Comment (kpouget) [inline recipes/simple-kmod/config/Makefile:6]: I think there is a mismatch between `config` and `manifests`, no?
```suggestion
	kubectl create configmap $@ -n $@ --from-file=recipes/$@/config/$@-cr.yaml
	-oc policy add-role-to-user system:image-puller system:serviceaccount:$@:builder -n driver-container-base
	@$(TEMPLATE_CMD) recipes/$@/manifests | kubectl apply -f - 
```

File: recipes/simple-kmod/manifests/0000-buildconfig..yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      ADD simple-kmod.conf .
+      RUN mkdir -p /usr/lib/kvc/
+      RUN mkdir -p /etc/kvc/
+      RUN make install
+
+  strategy:
+    dockerStrategy:
+      buildArgs:
+        - name: {{(index .SpecialResource.Spec.DriverContainer.BuildArgs 0).Name}}
+          value: {{(index .SpecialResource.Spec.DriverContainer.BuildArgs 0).Value}}
```
Comment (kpouget) [inline recipes/simple-kmod/manifests/0000-buildconfig..yaml]: it looks surprising that only index `0` is passed from `SpecialResource.Spec.DriverContainer.BuildArgs`, and not the other element. Is this on purpose?

File: recipes/simple-kmod/manifests/1000-driver-container.yaml [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ""
+      labels:
+        app: {{.SpecialResource.Name}}-{{.GroupName.DriverContainer}}-{{.OperatingSystemMajor}}
+    spec:
+      tolerations:
+      - operator: Exists
+      - key: nvidia.com/gpu
+        operator: Exists
+        effect: NoSchedule
```
Comment (kpouget) [inline recipes/simple-kmod/manifests/1000-driver-container.yaml]: I think that this recipe is not linked with the GPU driver, so this `NoSchedule` tolerations may not belong here

File: recipes/simple-kmod/config/Makefile (hunk 3 above)
Comment (kpouget) [reply recipes/simple-kmod/config/Makefile:6]: hum, forget about this comment, I studied further the code and now understand better how this works

File: recipes/simple-kmod/manifests/0000-buildconfig..yaml (hunk 4 above)
Comment (zvonkok) [reply recipes/simple-kmod/manifests/0000-buildconfig..yaml]: That is intentional, I am unsure if we can loop through templates, so this needs maybe some update. Good point.

File: recipes/simple-kmod/manifests/1000-driver-container.yaml (hunk 5 above)
Comment (zvonkok) [reply recipes/simple-kmod/manifests/1000-driver-container.yaml]: Yes you're right does not belong here.

Comment (zvonkok) [reply recipes/simple-kmod/manifests/1000-driver-container.yaml]:
```suggestion
        operator: Exists
        effect: NoSchedule
```

File: recipes/simple-kmod/config/simple-kmod-cr.yaml (hunk 2 above)
Comment (zvonkok) [reply recipes/simple-kmod/config/simple-kmod-cr.yaml]:
```suggestion
```

File: recipes/simple-kmod/manifests/0000-buildconfig..yaml (hunk 4 above)
Comment (kpouget) [reply recipes/simple-kmod/manifests/0000-buildconfig..yaml]: I didn't test it, but `range` seems to be what you need (see [this page](https://golang.org/pkg/text/template/)):
```suggestion
      buildArgs:
      {{ range .SpecialResource.Spec.DriverContainer.BuildArgs }}
        - name: {{ .Name }}
          value: {{ .Value }}
      {{ end }}
```

Comment (zvonkok) [reply recipes/simple-kmod/manifests/0000-buildconfig..yaml]: Will try that with the next commit.

Comment (zvonkok) [reply recipes/simple-kmod/manifests/0000-buildconfig..yaml]: This worked nicely, Updated the commit.

File: recipes/simple-kmod/manifests/1000-driver-container.yaml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          - name: overlay
+            mountPath: /tmp/overlay
+            readOnly: true
+      volumes:
+        - name: overlay
+          hostPath:
+            path: /var/lib/containers/storage/overlay
+        - name: run-{{.SpecialResource.Name}}
+          hostPath:
+            path: /run/nvidia
```
Comment (dagrayvid) [inline recipes/simple-kmod/manifests/1000-driver-container.yaml:147]: It seems like this volume and volumeMount are unneeded in this simple case. Is it just kept here as an example in the case where a driver container would need something like this?

**Pull request closed without merging**